### PR TITLE
✨♻️🏷️  254 - data mapper create participant

### PR DESCRIPTION
**Pull request by anncatton**
## Description of Changes

Implements `POST` `/participants` endpoint in Data Mapper for participant creation, as part of the registration flow.
Adds schema types to support these changes.

*Note: did not complete manual testing of the `POST` endpoint because there are changes required in both the PI and Consent DASes to support this flow. These will be implemented in subsequent tickets.

### Types

- moves guardianInfo check to utils file in `common` folder
- adds participant contact info (email, phoneNumber) check function. TBD whether these fields will be made optional when a guardian is present (#388 )
- adds modular schemas for Participant, in its own entity file
- updates Participant fields to match what is needed in registration flow
- renames `ParticipantIdentification` to `ParticipantIdentity`, and moves to its own entity file
- creates a separate `Guardian` entity file for code cleanup. This file includes a plain object for converting the optional fields being returned from the db
- adds request and response types to `services/data-mapper` for `CreateParticipant`, `ConsentCreateParticipant` (fields for consent-das) and `PICreateParticipant` (fields for pi-das)
- refactors `PIClinicianInvite` (`pi-das/responses`) and `ClinicianInvite` entity to use new Participant + Guardian sub schemas
- updates and cleans up `ClinicianInvite` and `ParticipantIdentity` tests
- changes regex imports to come directly from the `regexes.ts` file to fix an import resolution error
- adds `TODO`s for open questions being tracked in #388  

### Data Mapper

- moves existing DAS service calls to their own files
- updates `/participants` `POST` endpoint with request validation, error handling and jsdocs
- adds CreateParticipant schemas to swagger router
- updates router and service function parameters
- implements `createParticipant` service function that creates the participant in both PI and Consent DASes, and rolls back the changes if an error occurs.
- implements `deletePiParticipant` to support rollback in `createParticipant`
- implements separate service functions for creating participants in the PI and Consent DASes, with validation and error handling
- moves existing PHI and Keys DAS functions to their own `/services/das/` files. Functionality will be properly implemented in future tasks
- adds `TODO`s for open questions being tracked in #388 

## PR Readiness Checklist

- [x] "Expected Outcome(s)" in ticket have been met
- [x] Ticket number included in PR title
- [x] Connected ticket to PR
- [x] Labels added to PR for service name (`consent-api`, `data-mapper`, etc...), type (`chore`, `documentation`, etc...), status (`draft`, `on-hold`, etc...) **if applicable**
- [ ] Manual testing completed
- [x] Builds locally without errors or warnings
- [x] Tests are updated (if required) and passing
- [ ] PR feedback has been addressed **if applicable**
- [ ] New environment variables added to `.env.schema` files, `README.md`
- [x] Added copyrights to new files

### CODE REVIEW

File: apps/data-mapper/src/routers/participants.ts [hunk 1]
```diff
@@ -46,19 +49,59 @@ router.get('/:participantId', async (req, res) => {
...
+			switch (participant.status) {
+				case 'SUCCESS': {
+					return res.status(201).json(participant.data);
+				}
+				case 'PARTICIPANT_EXISTS': {
+					return res.status(409).json(ConflictErrorResponse(participant.message));
+				}
+				case 'SYSTEM_ERROR': {
+					return res.status(500).json(ServerErrorResponse(participant.message));
+				}
```
[the remark this answers is not in the document]
Comment (joneubank) [reply apps/data-mapper/src/routers/participants.ts:94]: Looks like we're missing a case for `InvalidRequest` which is a valid failure state for `CreateParticipantFailureStatus`. I suspect the solution is to remove `InvalidRequest` from that state union since it is never returned by `createParticipant`

File: apps/data-mapper/src/services/create.ts [hunk 2]
```diff
@@ -1,112 +1,123 @@
...
-// TODO: add Type instead of any
-const createParticipantOhipKey = async (participantId: string): Promise<any> => {
-	const { keysDasUrl } = getAppConfig();
-	// TODO: add error handling
-	const result = await axiosClient.post(urlJoin(keysDasUrl, 'ohip-keys'), {
-		participantId,
-	});
-	return result.data.ohipKey;
-};
+export type CreateParticipantFailureStatus = SystemError | InvalidRequest | 'PARTICIPANT_EXISTS';
```
Comment (joneubank) [inline apps/data-mapper/src/services/create.ts:22]: As mentioned in other comment, it looks like `InvalidReqeuest` is not needed here.

More importantly, these appear to be the failure status for the PI client request to create participant, so this type should be declared alongside that function. It can be reused here if desired. The problem right now is that `services/create.ts` is importing a function from `services/das/pi.ts`, while this type is being imported the otherway. This is a cycle in our imports. Cycles in imports is a "code smell" giving an indication that there is a logic problem in our code structure.

File: apps/data-mapper/src/services/create.ts [hunk 3]
```diff
@@ -1,112 +1,123 @@
...
+			currentLifecycleState,
+			emailVerified,
+		});
+
+		if (participantConsentData.status !== 'SUCCESS') {
+			// Unable to create participant in Consent DAS, rollback participant already created in PI-DAS
+			const deletePiParticipant = await deletePIParticipant(participantId);
+			if (deletePiParticipant.status !== 'SUCCESS') {
+				logger.error('Error deleting existing PI participant:', deletePiParticipant.message);
+				return failure('SYSTEM_ERROR', 'An unexpected error occurred.');
```
Comment (joneubank) [inline apps/data-mapper/src/services/create.ts:103]: We could probably improve this logging. This puts our data in an unexpected state, with the PI participant created but no consent data. It is probably recoverable but this state should be captured better in our log. Consider including participant ID so we know which ID to take action on if we are searching for this in our logs.

Comment (joneubank) [reply apps/data-mapper/src/services/create.ts:103]: As a next step after this PR... if our data is in this state, can the user resubmit their data successfully or will be wind up conflicting with "participant exists" type errors? Is action needed?

File: apps/data-mapper/src/services/das/consent.ts [hunk 4]
```diff
@@ -18,19 +18,21 @@
...
 } from 'types/dataMapper';
+import { ConsentCategory, ConsentQuestionArray, ConsentQuestionId } from 'types/entities';
+import { Result, failure, success } from 'types/httpResponses';
 import urlJoin from 'url-join';
 
 import { getAppConfig } from '../../config.js';
 import serviceLogger from '../../logger.js';
 import axiosClient from '../axiosClient.js';
-import { CreateInviteFailureStatus } from '../create.js';
+import { CreateInviteFailureStatus, CreateParticipantFailureStatus } from '../create.js';
```
Comment (joneubank) [inline apps/data-mapper/src/services/das/consent.ts:35]: As mentioned in comment above, the failure status type should probably be declared in this module along side `createParticipantConsentData`.

File: apps/data-mapper/src/services/das/keys.ts [hunk 5]
```diff
@@ -16,3 +16,19 @@
...
  */
+
+import urlJoin from 'url-join';
+
+import { getAppConfig } from '../../config.js';
+import axiosClient from '../axiosClient.js';
+
+// TODO: add proper JSDoc comments
+// TODO: add Type instead of any
+export const createParticipantOhipKey = async (participantId: string): Promise<any> => {
```
Comment (joneubank) [inline apps/data-mapper/src/services/das/keys.ts:27]: You're not asking me to approve a PR with a `Promise<any>` return type, are you? :P

Comment (joneubank) [reply apps/data-mapper/src/services/das/keys.ts:27]: lol TODO

File: apps/data-mapper/src/services/das/pi.ts [hunk 6]
```diff
@@ -136,3 +141,102 @@ export const deleteInvitePiData = async (
...
+			participantOhipLastName,
+			dateOfBirth,
+			participantPhoneNumber,
+			participantPreferredName,
+			participantEmailAddress,
+			guardianEmailAddress,
+			guardianPhoneNumber,
+			guardianRelationship,
+			keycloakId,
+		});
```
Comment (joneubank) [inline apps/data-mapper/src/services/das/pi.ts:179]: As far as i can tell, you never use these destructured properties individually. Why not just use `req` as an object instead of listing all the members?

File: apps/data-mapper/src/services/das/pi.ts [hunk 7]
```diff
@@ -136,3 +141,102 @@ export const deleteInvitePiData = async (
...
+			const participantId = data.participant.id;
+			if (participantId) {
+				// delete participant from PI DB if response parsing fails
+				logger.info('Deleting invalid participant');
+				const deletePiParticipant = await deletePIParticipant(participantId);
+				if (deletePiParticipant.status !== 'SUCCESS') {
+					logger.error('Error deleting existing PI participant:', deletePiParticipant.message);
+					return failure('SYSTEM_ERROR', 'An unexpected error occurred.');
+				}
+			}
```
Comment (joneubank) [inline apps/data-mapper/src/services/das/pi.ts:196]: This block surprises me. Are we preparing for a case where the request was successful but the response was invalid, so we create and then delete the participant? I actually think this might not be necessary. Let the PI service decide if creating the participant is correct. If the messaging fails to communicate the info about the participant then we handle this as a system error in everything downstream. Consider this similar to failing to fetch data, you don't request to delete data if the response is malformed.

File: apps/data-mapper/src/services/das/pi.ts [hunk 8]
```diff
@@ -136,3 +141,102 @@ export const deleteInvitePiData = async (
...
+	} catch (error) {
+		if (error instanceof AxiosError && error.response) {
+			const { data, status } = error.response;
+			logger.error('AxiosError handling delete participant request', data);
+
+			if (status === 404) {
+				return failure('PARTICIPANT_DOES_NOT_EXIST', data.message);
+			}
+
+			return failure('SYSTEM_ERROR', data.message);
```
Comment (joneubank) [inline apps/data-mapper/src/services/das/pi.ts:237]: 👍

File: packages/types/src/common/conditionalFieldUtils.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+	ADULT_CONSENT,
+	YOUNG_ADULT_CONSENT,
+} = ConsentGroup.enum;
+
+const requiresGuardianInformation = (consentGroup: ConsentGroup) => {
+	return (
+		consentGroup === GUARDIAN_CONSENT_OF_MINOR ||
+		consentGroup === GUARDIAN_CONSENT_OF_MINOR_INCLUDING_ASSENT ||
+		consentGroup === ADULT_CONSENT_SUBSTITUTE_DECISION_MAKER
+	);
```
Comment (joneubank) [inline packages/types/src/common/conditionalFieldUtils.ts:36]: Consider the following pattern to make this easier to extend in the future and prevent boolean logic typos.

```ts
const consentOptions: ConsentGroup[] = [GUARDIAN_CONSENT_OF_MINOR, GUARDIAN_CONSENT_OF_MINOR_INCLUDING_ASSENT, ADULT_CONSENT_SUBSTITUTE_DECISION_MAKER];

consentOptions.includes(consentGroup);
```

Comment (joneubank) [reply packages/types/src/common/conditionalFieldUtils.ts:36]: Note that the type of `consentOptions` needs to be `(typeof consentGroup)[]`, so in this case `ConsentGroup[]` so that the TS won't have issues with `consentOptions.includes(consentGroup)`.

File: packages/types/src/common/conditionalFieldUtils.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+const requiresParticipantContactInfo = (consentGroup: ConsentGroup) => {
+	return consentGroup === ADULT_CONSENT || consentGroup === YOUNG_ADULT_CONSENT;
+};
+
+const isUndefined = (arg: any): arg is undefined => arg === undefined;
+
+// TODO: decide if participant contact fields will be excluded in a guardian is present
+// TBD in https://github.com/OHCRN/platform/issues/388
+export const hasRequiredParticipantContactInfo = (
```
Comment (joneubank) [inline packages/types/src/common/conditionalFieldUtils.ts:47]: This function has mystery logic. A TSDoc comment would be helpful.

File: packages/types/src/common/conditionalFieldUtils.ts [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+};
+
+export const hasRequiredGuardianInformation = (
+	props: {
+		consentGroup: ConsentGroup;
+	} & GuardianBaseFields,
+) => {
+	// guardianName, guardianPhoneNumber, guardianEmailAddress, guardianRelationship must be defined if
+	// GUARDIAN_CONSENT_OF_MINOR, GUARDIAN_CONSENT_OF_MINOR_INCLUDING_ASSENT or ADULT_CONSENT_SUBSTITUTE_DECISION_MAKER was selected
+
```
Comment (joneubank) [inline packages/types/src/common/conditionalFieldUtils.ts:65]: Looks like a TSDoc description to me!

File: packages/types/src/common/conditionalFieldUtils.ts [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+		consentGroup,
+		guardianName,
+		guardianPhoneNumber,
+		guardianEmailAddress,
+		guardianRelationship,
+	} = props;
+
+	return requiresGuardianInformation(consentGroup)
+		? [guardianName, guardianPhoneNumber, guardianEmailAddress, guardianRelationship].every(
+				(field) => !isUndefined(field),
```
Comment (joneubank) [inline packages/types/src/common/conditionalFieldUtils.ts:76]:
```ts
const hasValue = <T>(input: T | undefined): input is T => !isUndefined(input);

// ...

someArray.every(hasValue);
```

File: packages/types/tests/entities/ClinicianInvite.spec.ts [hunk 13]
```diff
@@ -26,86 +26,82 @@ import { PIClinicianInviteResponse } from '../../src/services/piDas/index.js';
...
-				participantLastName: 'Simpson',
-				participantEmailAddress: 'bart.simpson@example.com',
-				participantPhoneNumber: '6471234567',
-				consentGroup: ConsentGroup.enum.GUARDIAN_CONSENT_OF_MINOR,
-				guardianName: 'Marge Simpson',
-				guardianRelationship: 'Wife', // missing guardianEmailAddress and guardianPhoneNumber
-				consentToBeContacted: true,
-			}).success,
-		).false;
+		const result = ClinicianInviteBase.safeParse({
```
Comment (joneubank) [inline packages/types/tests/entities/ClinicianInvite.spec.ts:71]: This is a good use of test over zod schema, making sure a specific refinement is being enforced.

Perhaps a test fixture on the common refinement functions themselves would be useful. Maybe that already exists...

File: packages/types/tests/entities/ParticipantIdentity.spec.ts [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+			ohipNumber: '1234567890',
+			currentLifecycleState: LifecycleState.enum.REGISTERED,
+			participantPreferredName: 'Homer',
+			participantOhipFirstName: 'Homer',
+			participantOhipLastName: 'Simson',
+			dateOfBirth: new Date(),
+			participantPhoneNumber: '6471234567',
+			residentialPostalCode: 'T4B0V7',
+			participantEmailAddress: 'homer.simpson@example.com',
+			consentGroup: ConsentGroup.enum.GUARDIAN_CONSENT_OF_MINOR, // missing all guardian contact fields
```
Comment (joneubank) [inline packages/types/tests/entities/ParticipantIdentity.spec.ts:87]: These comments are incredibly helpful. Thank you.

**Review by joneubank — COMMENTED** [review]
Loving the code patterns here in general. Was pretty clear to read through the work. I have left you a bunch of thoughts for places to do a bit of improvement.

File: apps/data-mapper/src/services/das/consent.ts [hunk 15]
```diff
@@ -212,3 +214,33 @@ export const getParticipantResponsesByQuestionId = async ({
...
+	try {
+		const result = await axiosClient.post(urlJoin(consentDasUrl, 'participants'), req);
+		return success(result.data.participant);
+	} catch (error) {
+		if (error instanceof AxiosError && error.response) {
+			const { data, status } = error.response;
+
+			if (status === 400) {
+				logger.error('Invalid request while creating participant', data);
+				return failure('INVALID_REQUEST', data.message);
```
Comment (mistryrn) [inline apps/data-mapper/src/services/das/consent.ts:236]: Looks like this may be the case where `INVALID_REQUEST` is being returned by the `createParticipant` function, so we may need to add a case to the switch statement after all

File: apps/data-mapper/src/services/das/keys.ts (hunk 5 above)
Comment (mistryrn) [reply apps/data-mapper/src/services/das/keys.ts:27]: I can clarify these @joneubank  😅 They're from the original Data Mapper POC many months ago and Ann is just moving my hasty POC code around here. We haven't properly implemented these yet, they will get addressed in the coming sprints when we work on this functionality

TL;DR: it's all my fault, but we'll be fixing it soon 🙈

File: packages/types/src/common/conditionalFieldUtils.ts (hunk 9 above)
Comment (anncatton) [reply packages/types/src/common/conditionalFieldUtils.ts:36]: aha that is prob what i was missing! i had tried `includes` and got TS yelling, but admittedly got lazy and just moved on

File: packages/types/tests/entities/ClinicianInvite.spec.ts (hunk 13 above)
Comment (anncatton) [reply packages/types/tests/entities/ClinicianInvite.spec.ts:71]: i think at the moment our tests our only with the refinement applied to specific schemas using it. Will try this out.

File: packages/types/tests/entities/ParticipantIdentity.spec.ts (hunk 14 above)
Comment (anncatton) [reply packages/types/tests/entities/ParticipantIdentity.spec.ts:87]: i think the credit here goes to Alyssa 🙂

File: apps/data-mapper/src/services/das/keys.ts (hunk 5 above)
Comment (anncatton) [reply apps/data-mapper/src/services/das/keys.ts:27]: in my mind i had put a note about this in the PR description - but i DIDN'T!! my bad. but yes, these will be implemented properly when the services are. I just wanted to get these functions into the proper files for now.

File: apps/data-mapper/src/services/das/pi.ts (hunk 7 above)
Comment (anncatton) [reply apps/data-mapper/src/services/das/pi.ts:196]: I'm also not sure it's necessary - the reason i thought of this roll back at all is that PI DAS schema is not up to date with the schema types changes i have in this pr, so testing locally the pi das response was automatically invalid. (I've a ticket created to implement the field changes in the DAS: https://github.com/OHCRN/platform/issues/391 )

Just to make sure I understand your suggestion - we rely on the PI DAS itself to validate its response data (and rollback if necessary), so that Data Mapper wouldn't be responsible for that portion of the transaction?

We do have this pattern elsewhere in this PR (to delete pi data if consent participant step fails) and in the Clinician Invite flow. Sounds like maybe we should revisit?

**Review by mistryrn — COMMENTED** [review]
Lots of great stuff here, really liking the new types structure/organization!

File: apps/data-mapper/src/services/das/consent.ts (hunk 15 above)
Comment (anncatton) [reply apps/data-mapper/src/services/das/consent.ts:236]: i think what i have here is incorrect - i forgot to handle the conflict error instead, so this should be checking for a `409` from the consent das (like i have in the pi create service call). The `withRequestValidation` wrapper on the router `POST` will return a `400` if the request body doesn't validate against the `CreateParticipantRequest` schema, so we shouldn't need the explicit case as I have it here.

File: packages/types/src/entities/ParticipantIdentity.ts
[the remark this answers is not in the document]
Comment (anncatton) [reply packages/types/src/entities/ParticipantIdentity.ts:1]: yeah maybe i can bring this up in our demo meeting!

File: apps/data-mapper/src/routers/participants.ts [hunk 16]
```diff
@@ -46,19 +49,59 @@ router.get('/:participantId', async (req, res) => {
...
+ *             $ref: '#/components/schemas/CreateParticipantRequest'
+ *     responses:
+ *       201:
+ *         description: OK
+ *         content:
+ *           application/json:
+ *             schema:
+ *               $ref: '#/components/schemas/CreateParticipantResponse'
+ *       400:
+ *         description: RequestValidationError - The request body was invalid.
```
[the remark this answers is not in the document]
Comment (anncatton) [reply apps/data-mapper/src/routers/participants.ts:74]: we don't, but sounds like a nice idea!
I will add an enhancement ticket for this to avoid making this PR larger than it needs to be

[the remark this answers is not in the document]
Comment (anncatton) [reply apps/data-mapper/src/routers/participants.ts:74]: https://github.com/OHCRN/platform/issues/399

File: packages/types/src/common/conditionalFieldUtils.ts [hunk 17]
```diff
@@ -29,40 +29,65 @@ const {
...
 // TBD in https://github.com/OHCRN/platform/issues/388
+/**
+ * Checks if a Participant schema object contains the required Participant contact fields needed for the ConsentGroup
+ *
+ * participantEmailAddress and participantPhoneNumber must be defined if ADULT_CONSENT or YOUNG_ADULT_CONSENT was selected
+ * @param {Object} props
+ * @param {ConsentGroup} props.consentGroup - the ConsentGroup associated with a Participant object
+ * @param {string} props.participantEmailAddress - value for participantEmailAddress in a Participant object
+ * @param {string} props.participantPhoneNumber - value for participantPhoneNumber in a Participant object
+ * @returns {boolean} returns true if all required fields are present
```
Comment (joneubank) [inline packages/types/src/common/conditionalFieldUtils.ts:58]: To help simplify... I don't think we need to declare the types of variables in the TSDocs. They are prone to becoming out of sync with the actual code, and if that happens I'll trust the TS interpretter over the docs anyways. The descriptive text is really the win here.

Comment (joneubank) [reply packages/types/src/common/conditionalFieldUtils.ts:58]: ... In this case, we might not even need the description of the params. We know what a consentGroup.participantPhoneNumber is I think.

Comment (joneubank) [reply packages/types/src/common/conditionalFieldUtils.ts:58]: The return description (true if all fields are present) is real helpful though

File: packages/types/src/common/conditionalFieldUtils.ts [hunk 18]
```diff
@@ -29,40 +29,65 @@ const {
...
 };
 
 const requiresParticipantContactInfo = (consentGroup: ConsentGroup) => {
-	return consentGroup === ADULT_CONSENT || consentGroup === YOUNG_ADULT_CONSENT;
+	const consentOptions: ConsentGroup[] = [ADULT_CONSENT, YOUNG_ADULT_CONSENT];
+	return consentOptions.includes(consentGroup);
 };
 
 const isUndefined = (arg: any): arg is undefined => arg === undefined;
+const hasValue = <T>(input: T | undefined): input is T => !isUndefined(input);
```
Comment (joneubank) [inline packages/types/src/common/conditionalFieldUtils.ts]: There's no right answer to this question but: are these the kind of functions we are likely to want to reuse? Should they be moved somewhere to make them available in other files? other modules? How many instances of these functions are future devs going to make? Maybe thats ok?

File: packages/types/tests/utils/conditionalFields.spec.ts [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+		YOUNG_ADULT_CONSENT, // non-guardian group
+	} = ConsentGroup.enum;
+	describe('hasRequiredGuardianInformation', () => {
+		it('returns TRUE for a guardian consentGroup (GUARDIAN_CONSENT_OF_MINOR) and all required guardian fields are provided', () => {
+			const guardianFields = {
+				guardianName: 'Gina Guardian',
+				guardianEmailAddress: 'gina_g@example.com',
+				guardianPhoneNumber: '0123456789',
+				guardianRelationship: 'Mother',
+			};
```
Comment (joneubank) [inline packages/types/tests/utils/conditionalFields.spec.ts:43]: This object is repeated a lot, you could define once at top of the file and reuse it.

File: packages/types/tests/utils/conditionalFields.spec.ts [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+			expect(result).false;
+		});
+
+		it('returns TRUE for a non-guardian consentGroup and no guardian fields are provided', () => {
+			const testSchemaObj = {
+				consentGroup: ADULT_CONSENT,
+			};
+			const result = hasRequiredGuardianInformation(testSchemaObj);
+			expect(result).true;
+		});
```
Comment (joneubank) [inline packages/types/tests/utils/conditionalFields.spec.ts:122]: Is there an inverse case for this? What do we return for a non-guardian when consentGroups ARE provided.

**Review by joneubank — DISMISSED** [review]
Looks great Ann, thanks.

I know it's not always possible but breaking this into smaller PRs is a good objective for us all.

File: packages/types/src/common/conditionalFieldUtils.ts (hunk 17 above)
Comment (anncatton) [reply packages/types/src/common/conditionalFieldUtils.ts:58]: yeah the other thing about that is that Intellisense doesn't actually show the type in the hover definition, if following the `{type}` syntax. Not sure if this is a problem with VSCode or jsdoc!

Comment (anncatton) [reply packages/types/src/common/conditionalFieldUtils.ts:58]: i made a little compromise and left in `@params props prop1, prop2, ...`

**Review by mistryrn — APPROVED** [review]
🚀
